File: application/app/src/main/java/com/yahyeet/boardbook/model/handler/populator/GameTeamPopulator.java
package com.yahyeet.boardbook.model.handler.populator;

import com.yahyeet.boardbook.model.entity.Game;
import com.yahyeet.boardbook.model.entity.GameRole;
import com.yahyeet.boardbook.model.entity.GameTeam;
import com.yahyeet.boardbook.model.repository.IGameRepository;
import com.yahyeet.boardbook.model.repository.IGameRoleRepository;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class GameTeamPopulator {
	IGameRepository gameRepository;
	IGameRoleRepository gameRoleRepository;

	public GameTeamPopulator(IGameRepository gameRepository, IGameRoleRepository gameRoleRepository) {
		this.gameRepository = gameRepository;
		this.gameRoleRepository = gameRoleRepository;
	}

	public CompletableFuture<GameTeam> populate(GameTeam gameTeam) {
		GameTeam populatedGameTeam = new GameTeam(gameTeam.getName());
		populatedGameTeam.setId(gameTeam.getId());

		CompletableFuture<Game> futureGame = gameRepository.find(gameTeam.getGame().getId());
		CompletableFuture<List<GameRole>> futureGameRoles = gameRoleRepository.findRolesByTeamId(populatedGameTeam.getId());

		return futureGame.thenCombine(futureGameRoles, (game, roles) -> {
			populatedGameTeam.setGame(game);
			roles.forEach(populatedGameTeam::addRole);

			return populatedGameTeam;
		});
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/handler/populator/GameTeamPopulator.kt
package com.yahyeet.boardbook.model.handler.populator

import com.yahyeet.boardbook.model.entity.Game
import com.yahyeet.boardbook.model.entity.GameRole
import com.yahyeet.boardbook.model.entity.GameTeam
import com.yahyeet.boardbook.model.repository.IGameRepository
import com.yahyeet.boardbook.model.repository.IGameRoleRepository
import java.util.concurrent.CompletableFuture

class GameTeamPopulator(private val gameRepository: IGameRepository, private val gameRoleRepository: IGameRoleRepository) {

    fun populate(gameTeam: GameTeam): CompletableFuture<GameTeam> {
        val populatedGameTeam = GameTeam(gameTeam.name)
        populatedGameTeam.id = gameTeam.id

        val futureGame = gameRepository.find(gameTeam.game!!.id)
        val futureGameRoles = gameRoleRepository.findRolesByTeamId(populatedGameTeam.id)

        return futureGame.thenCombine(futureGameRoles) { game, roles ->
            populatedGameTeam.game = game
            roles.forEach(Consumer<GameRole> { populatedGameTeam.addRole(it) })

            populatedGameTeam
        }
    }
}
